server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1016,15 +1066,103 @@ public void writeTo(StreamOutput out) throws IOException {
                 buildUnifiedCustomDiff().writeTo(out);
                 buildUnifiedReservedStateMetadataDiff().writeTo(out);
             } else {
-                clusterCustoms.writeTo(out);
-                reservedStateMetadata.writeTo(out);
-                if (multiProject != null) {
-                    multiProject.writeTo(out);
+                final var multiProjectToWrite = multiProject != null
+                    ? multiProject
+                    : DiffableUtils.singleEntryDiff(DEFAULT_PROJECT_ID, singleProject, PROJECT_ID_SERIALIZER);
+
+                if (out.getTransportVersion().before(TransportVersions.REPOSITORIES_METADATA_AS_PROJECT_CUSTOM)) {
+                    writeDiffWithRepositoriesMetadataAsClusterCustom(out, clusterCustoms, multiProjectToWrite, reservedStateMetadata);
                 } else {
-                    // construct the MapDiff to write out this single project
-                    DiffableUtils.singleEntryDiff(DEFAULT_PROJECT_ID, singleProject, PROJECT_ID_SERIALIZER).writeTo(out);
+                    clusterCustoms.writeTo(out);
+                    reservedStateMetadata.writeTo(out);
+                    multiProjectToWrite.writeTo(out);
+                }
+            }
+        }
+
+        @SuppressWarnings({ "rawtypes", "unchecked" })
+        private static void writeDiffWithRepositoriesMetadataAsClusterCustom(
+            StreamOutput out,
+            MapDiff<String, ClusterCustom, ImmutableOpenMap<String, ClusterCustom>> clusterCustoms,
+            MapDiff<ProjectId, ProjectMetadata, Map<ProjectId, ProjectMetadata>> multiProject,
+            MapDiff<String, ReservedStateMetadata, ImmutableOpenMap<String, ReservedStateMetadata>> reservedStateMetadata
+        ) throws IOException {
+            assert out.getTransportVersion().onOrAfter(TransportVersions.MULTI_PROJECT)
+                && out.getTransportVersion().before(TransportVersions.REPOSITORIES_METADATA_AS_PROJECT_CUSTOM) : out.getTransportVersion();
```

Comment:
@DaveCTurner I meant this method which is most of the new complexity. Versions between `MULTI_PROJECT` and `REPOSITORIES_METADATA_AS_PROJECT_CUSTOM` are only released in serverless. Once serverless rolls forward, we can remove it.